File: Scripts/volatile_and_force_compiler_task/numerical-integration/c++/numerical-integration.cpp
// the integration routine

#include <iostream>

double aVar=0;

template<typename Method, typename F, typename Float>
 double integrate(F f, Float a, Float b, int steps, Method m)
{
  double s = 0;
  double h = (b-a)/steps;
  for (int i = 0; i < steps; ++i)
    s += m(f, a + h*i, h);
  return h*s;
}

// methods
class rectangular
{
public:
  enum position_type { left, middle, right };
  rectangular(position_type pos): position(pos) {}
  template<typename F, typename Float>
   double operator()(F f, Float x, Float h) const
  {
    switch(position)
    {
    case left:
      return f(x);
    case middle:
      return f(x+h/2);
    case right:
      return f(x+h);
    }
  }
private:
  const position_type position;
};

class trapezium
{
public:
  template<typename F, typename Float>
   double operator()(F f, Float x, Float h) const
  {
    return (f(x) + f(x+h))/2;
  }
};

class simpson
{
public:
  template<typename F, typename Float>
   double operator()(F f, Float x, Float h) const
  {
    return (f(x) + 4*f(x+h/2) + f(x+h))/6;
  }
};

// sample usage
double f1(double x) { return (x*x*x)+aVar; }
double f2(double x) { return 1/(x+aVar); }
double f3(double x) { return x+aVar; }

int executeTask(int r) {
r = integrate(f1, 0.0, 1.0, 100, rectangular(rectangular::left));
r = integrate(f1, 0.0, 1.0, 100, rectangular(rectangular::middle));
r = integrate(f1, 0.0, 1.0, 100, rectangular(rectangular::right));
r = integrate(f1, 0.0, 1.0, 100, trapezium());
r = integrate(f1, 0.0, 1.0, 100, simpson());

r = integrate(f2, 1.0, 100.0, 1000, rectangular(rectangular::left));
r = integrate(f2, 1.0, 100.0, 1000, rectangular(rectangular::middle));
r = integrate(f2, 1.0, 100.0, 1000, rectangular(rectangular::right));
r = integrate(f2, 1.0, 100.0, 1000, trapezium());
r = integrate(f2, 1.0, 100.0, 1000, simpson());

r = integrate(f3, 0.0, 5000.0, 5000000, rectangular(rectangular::left));
r = integrate(f3, 0.0, 5000.0, 5000000, rectangular(rectangular::middle));
r = integrate(f3, 0.0, 5000.0, 5000000, rectangular(rectangular::right));
r = integrate(f3, 0.0, 5000.0, 5000000, trapezium());
r = integrate(f3, 0.0, 5000.0, 5000000, simpson());

r = integrate(f3, 0.0, 6000.0, 6000000, rectangular(rectangular::left));
r = integrate(f3, 0.0, 6000.0, 6000000, rectangular(rectangular::middle));
r = integrate(f3, 0.0, 6000.0, 6000000, rectangular(rectangular::right));
r = integrate(f3, 0.0, 6000.0, 6000000, trapezium());
r = integrate(f3, 0.0, 6000.0, 6000000, simpson());
	

	return 1 + r;
}

int main() {

	volatile double r;
	for ( aVar = 0; aVar < 100; ++aVar ) {
		r = executeTask(aVar);
	}

return 0;
}

File: Scripts/volatile_and_force_compiler_task/numerical-integration/r/numerical-integration.r
aVar=0

integrate.rect <- function(f, a, b, n, k=0) {
  #k = 0 for left, 1 for right, 0.5 for midpoint
  h <- (b-a)/n
  x <- seq(a, b, len=n+1)
  sum(f(x[-1]-h*(1-k)))*h
}

integrate.trapezoid <- function(f, a, b, n) {
  h <- (b-a)/n
  x <- seq(a, b, len=n+1)
  fx <- f(x)
  sum(fx[-1] + fx[-length(x)])*h/2 
}

integrate.simpsons <- function(f, a, b, n) {
  h <- (b-a)/n
  x <- seq(a, b, len=n+1)
  fx <- f(x)
  sum(fx[-length(x)] + 4*f(x[-1]-h/2) + fx[-1]) * h/6
}

f1 <- (function(x) {(x^3)+aVar})
f2 <- (function(x) {1/(x+aVar)})
f3 <- (function(x) {x+aVar})
f4 <- (function(x) {x+aVar})

executeTask <- function(i) {

  integrate.rect(f1,0,1,100,0) #TopLeft 0.245025
  integrate.rect(f1,0,1,100,0.5) #Mid 0.2499875
  integrate.rect(f1,0,1,100,1) #TopRight 0.255025
  integrate.trapezoid(f1,0,1,100) # 0.250025
  integrate.simpsons(f1,0,1,100) #0.25

  integrate.rect(f2,1,100,1000,0) #TopLeft 0.245025
  integrate.rect(f2,1,100,1000,0.5) #Mid 0.2499875
  integrate.rect(f2,1,100,1000,1) #TopRight 0.255025
  integrate.trapezoid(f2,1,100,1000) # 0.250025
  integrate.simpsons(f2,1,100,1000) #0.25

  integrate.rect(f3,0,5000,5000000,0) #TopLeft 0.245025
  integrate.rect(f3,0,5000,5000000,0.5) #Mid 0.2499875
  integrate.rect(f3,0,5000,5000000,1) #TopRight 0.255025
  integrate.trapezoid(f3,0,5000,5000000) # 0.250025
  integrate.simpsons(f3,0,5000,5000000) #0.25

  integrate.rect(f3,0,6000,6000000,0) #TopLeft 0.245025
  integrate.rect(f3,0,6000,6000000,0.5) #Mid 0.2499875
  integrate.rect(f3,0,6000,6000000,1) #TopRight 0.255025
  integrate.trapezoid(f3,0,6000,6000000) # 0.250025
  integrate.simpsons(f3,0,6000,6000000) #0.25
	return(i+1)
}

r=1

for (aVar in 0:100) {
	r = executeTask(aVar)
}
